src/crepe/system/AISystem.cpp

```diff
@@ -0,0 +1,194 @@
+		vec2 force_to_add = this->flee(ai, rigidbody);
+
+		if (!this->accumulate_force(ai, force, force_to_add)) {
+			return force;
+		}
+	}
+	if (ai.on(AI::BehaviorType::ARRIVE)) {
+		vec2 force_to_add = this->arrive(ai, rigidbody);
+
+		if (!this->accumulate_force(ai, force, force_to_add)) {
+			return force;
+		}
+	}
+	if (ai.on(AI::BehaviorType::SEEK)) {
+		vec2 force_to_add = this->seek(ai, rigidbody);
+
+		if (!this->accumulate_force(ai, force, force_to_add)) {
+			return force;
+		}
+	}
+	if (ai.on(AI::BehaviorType::PATH_FOLLOW)) {
+		vec2 force_to_add = this->path_follow(ai, rigidbody);
+
+		if (!this->accumulate_force(ai, force, force_to_add)) {
+			return force;
+		}
+	}
+
+	return force;
+}
+
+bool AISystem::accumulate_force(const AI & ai, vec2 & running_total, vec2 & force_to_add) {
+	float magnitude = running_total.length();
+	float magnitude_remaining = ai.max_force - magnitude;
+
+	if (magnitude_remaining <= 0.0f) {
+		// If the force is already at/above the max force, return false
+		return false;
+	}
+
+	float magnitude_to_add = force_to_add.length();
+	if (magnitude_to_add < magnitude_remaining) {
+		// If the force to add is less than the remaining force, add it
+		running_total += force_to_add;
+	} else {
+		// If the force to add is greater than the remaining force, add a fraction of it
+		force_to_add.normalize();
+		running_total += force_to_add * magnitude_remaining;
+	}
+
+	return true;
+}
+
+vec2 AISystem::seek(const AI & ai, const Rigidbody & rigidbody) const {
+	const Mediator & mediator = this->mediator;
+	ComponentManager & mgr = mediator.component_manager;
+	RefVector<Transform> transforms = mgr.get_components_by_id<Transform>(ai.game_object_id);
+	Transform & transform = transforms.front().get();
+
+	// Calculate the desired velocity
+	vec2 desired_velocity = ai.seek_target - transform.position;
+	desired_velocity.normalize();
+	desired_velocity *= rigidbody.data.max_linear_velocity;
+
+	return desired_velocity - rigidbody.data.linear_velocity;
+}
+
+vec2 AISystem::flee(const AI & ai, const Rigidbody & rigidbody) const {
+	const Mediator & mediator = this->mediator;
+	ComponentManager & mgr = mediator.component_manager;
+	RefVector<Transform> transforms = mgr.get_components_by_id<Transform>(ai.game_object_id);
+	Transform & transform = transforms.front().get();
+
+	// Calculate the desired velocity if the entity is within the panic distance
+	vec2 desired_velocity = transform.position - ai.flee_target;
+	if (desired_velocity.length_squared() > ai.square_flee_panic_distance) {
+		return vec2{0, 0};
+	}
+	desired_velocity.normalize();
+	desired_velocity *= rigidbody.data.max_linear_velocity;
+
+	return desired_velocity - rigidbody.data.linear_velocity;
+}
+
+vec2 AISystem::arrive(const AI & ai, const Rigidbody & rigidbody) const {
+	const Mediator & mediator = this->mediator;
+	ComponentManager & mgr = mediator.component_manager;
+	RefVector<Transform> transforms = mgr.get_components_by_id<Transform>(ai.game_object_id);
```

Comment:
misschien ook de transform meegeven net als de rigidbody

Reply:
aangezien er vaak ai,rigidbody,transform nodig is kan hier een struct voor gemaakt worden die ze samenvoegd en deze doorgestuurd worden

Reply:
Ik zal transform ook doorgeven by reference, maar het samenvoegen in één struct is lastiger omdat AI niet overal const kan zijn (dus dan heb ik al twee structs nodig). Is het oke om geen struct te gebruiken?

Reply:
ja hoor dat is niet erg